mib.1
\."#################################################################
\."$Id: mib.1 663 2003-12-19 16:39:10Z garlick $
\."by Andrew C. Uselton <uselton2@llnl.gov> 
\."#################################################################
\."  Copyright (C) 2006 The Regents of the University of California.
\."  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
\."  Written by Andrew Uselton (uselton2@llnl.gov>
\."  UCRL-CODE-222725
\."  
\."  This file is part of Mib, a remote power management program.
\."  For details, see <http://www.llnl.gov/linux/mib/>.
\."  
\."  Mib is free software; you can redistribute it and/or modify it under
\."  the terms of the GNU General Public License (as published by the Free
\."  Software Foundation) version 2, dated June 1991.
\."  
\."  Mib is distributed in the hope that it will be useful, but WITHOUT 
\."  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
\."  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
\."  for more details.
\."  
\."  You should have received a copy of the GNU General Public License along
\."  with Mib; if not, write to the Free Software Foundation, Inc.,
\."  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\."#################################################################
.\"
.TH Mib 1 "Release 1.9.6" "LLNL" "Mib"
.SH NAME
mib \- An MPI I/O Benchmark
.SH SYNOPSIS
.B mib
.I "[-option]..."
.SH DESCRIPTION
Perform I/O and report on the rates observed.
.SH OPTIONS
.TP
.I "-b [<granularity>], --random_reads [<granularity>]"
Perform a seek to a random location in the file before each read.  The optional
ganularity constrains the targets of the seek to offsets in the file equal to
multiples of that value (default no seeks).  
.TP
.I "-E, --show_environment"
Show environment of test in output (default false).
.TP
.I "-F, --force"
By default,
.B mib
will not perform I/O to any file system type (as returned by statfs in the f_type 
field) but Lustre.  This is a safety feature.  Sending benchmark I/O, especially 
inadvertently, to an NFS mounted home directory can amount to a denial of service
attack and make the perpetrator very unpopular.  If that is really what is wanted
then this option will turn off the safety check.
.TP
.I "-I, --show_intermediate_values"
Show intermediate values in output (default false).
.TP
.I "-h, --help"
Print a help message and exit.
.TP
.I "-H, --show_headers"
Show headers for the fields in the results line (default false).
.TP
.I "-l <count>, --call_limit <count>"
Issue no more than <count> system calls (default 4096).  The limit is separately 
enforced for the write phase and the read phase.
.TP
.I "-L <seconds>, --time_limit <seconds>"
Do not issue new system calls after <seconds> seconds. Again, limits are per phase 
for write and read phases (default 60 seconds).
.TP
.I "-M, --no_mpi "
.B mib
is usually run under a resource manager (eg. SLURM) that sets up an MPI context.  
It will run without, though, either locally or via a remote launcher (eg. pdsh).
.B mib
detects if MPI is available and uses it if it is.  If MPI is available and one wants
to test the non-MPI behavior this switch will do that.
.TP
.I "-n, --new "
Create new files if files were already present.  It will always create new files 
if none were present (default false, i.e. reuse any files that already exist).
.TP
.I "-p <profile>, --profiles <profile>"
Output system call timing profiles to 
.I <profile>.write
and 
.I <profile>.read 
(default off).  A profile is a text file with one line per system call issued and 
one column per task.  Each entry is the MPI global clock reading for the time when
the given system call returned in the given task.  All timing values are relative 
to a barrier-coordinated moment at the beginning of the test.  Since some tasks may
issue more system calls than others the remaining entries are set to zero.  At this 
time, 
.B mib
will not generate profile files without MPI.
.TP
.I "-P, --show_progress"
Show progress bars during testing (default false).  This just prints a series of 
asterisk characters in a best effort to show the progress of the test.  Due to 
delays in the system calls and buffering from the remote tasks the bars may or 
may not follow the progress closely.
.TP
.I "-r, --remove"
Remove files when done (default false).
.TP
.I "-R, --read_only"
Only perform the read test (default false).
.TP
.I "-s <size>, --call_size <size>"
Use system calls of this size (default 512k).  Numbers may use abreviations 
k, K, m, or M.  Both k and K are interpreted as 1024, and m and M are k*k.
.TP
.I "-S, --show_signon"
Show signon message, including date, program name, and version number.
.TP
.I "-t <dir>, --test_dir <dir>"
Required. I/O transactions to and from files in <dir>.  If a task number is 
available from MPI or SLURM that value is used as the <index> in the target 
file name for the task as follows: <dir>/mibData.<index>.  If no <index> can
be discerned an attempt is made to use the hostname in a similar fashion.  For
writes, the <index> directly corresponds to the task number.  For reads, a 
formula attempts to make the read target as distant as possible from the 
write target.  This makes the reads exercise lock acquisition and release and
cache flushing.
.TP
.I "-V, --version"
Print the version and exit.
.TP
.I "-W, --write_only"
Only perform the write test.  If -R and -W are both present both tests 
will run, but that's the default anyway (default false).

.SH OPERATION
When invoked with a context enabling MPI operations (via 
.I srun
for intance)
.B mib
writes and the reads files, one per task, reports on the 
observed data rates, and optionally
reports on the timings of the system calls.  For details refer to the 
discussion in 
.I /usr/share/doc/mib-<version>/mib.pdf.

.SH FILES
<profile>.write
.br
<profile>.read
.SH ORIGIN
Developed by Andrew  Uselton <useton2@llnl.gov> on LLNL's Linux 
clusters.  This software is open source and distributed under
the terms of the Gnu GPL.  



ChangeLog
2006-07-13 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.6

	* Update all files with UCRL number in preparation for release.

	* Sanity check documentation.
	
2006-06-29 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.5

	* Modify dlopen() code to be configurable, set up elan and
	infiniband  variants.

	* Cast the "size_t count" parameter to Write() and Read() to be
	ssize_t.  This prevents incorrect execution when the
	system call returns an error condition (-1) and misinterprets it
	as a very large number.  There is also one in Snprintf.

	* Include "-Wextra" as a CCFLAGS flag to catch future instances of
	the above.  
	
2006-04-26 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.4

	* Add lwatch.py to the project, since lmtd actually is available
	as open source.  Modify references saying otherwise in the 
	documentations.
	
2006-04-20 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.3

	* Fix really stupid, pernicious, and fundamental bug regarding use
	of opt->flags and verbosity references in macros.  They appear in
	several places, especially base_report, before opts is
	initialized.  1) split base_report from flags test and introduce
	new conditional_report.  2) do not refer to opts in macros defined
	in options.h.  Use opts in macros only locally to mib.c.
	
2006-04-3 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.2

	* Take out the node averaging.  The BGL code has entirely forked
	and node averaging is only an issue there.  It is also less reasonable
	on non-BGL clusters because the tasks do not stay coordinated across 
	a node like they do on BGL.

	* The -V option should include the architecture, not just the signon 
	message.

	* regularize all the documentation so it all agrees.
	
2006-03-24 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9.1

	* Add the architecture to the version string

2006-03-17 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.9

	* Allow for operation without MPI

	* Introduce random read option

	* Improve the mib.sh wrapper script

	* Make the target a required argument to avoid accidentally
	writing to NFS and causing disruption

	* Check for the file system type of the target and refuse to use
	NFS unless a "-F" (--force) flag is set.

	* Add a version option that prints the version string and exits.

	* Remove the "options" file code, use command line arguments only

	* Add a unit test "mib-test.sh" to the tools directory

	* Fix Mib_Init, it didn't return(mib) its result

	* When there is no SLURM or MPI we set the rank and base to the 
	host's index.  That means the read test's file name needs the 
	base added to its modulo arithmatic.  When we have arbitrary 
	subcommunicators, if ever, that would also come up.

	* "Open" for writes should have O_CREAT, and "Open" for reads
	should not.  Doing this makes the "Open" implementation straight
	forward.  

	* If the statfs call fails, which it will if you gave it a bogus
	target directory, then you need to bail with an error message.
	I'd left out the "exit(1)".

	* Convert to using config.h, which is created by configure, to
	detect if MPI is present on the build host, and skip using it if
	so.

	* Set the correct include directory: /usr/lib/mpi/include in the
	Makefile.in
	
2006-03-01 Andrew Uselton <uselton2@llnl.gov>
	
	* Tag 1.8

	* Dump the iterateions and pause options

	* Code factoring in the options settings

	* Consolidate boolean flags into fields in a single flag

	* Fixed ENOENT / ENONET bug :(

	* Get command line options (including long-opts) working and tested

	* Allow there to be a log_dir with no "options" file

	* Make a man page

	* Fix packaging

	
2006-02-15 Andrew Uselton <uselton2@llnl.gov>
	
	* Tag 1.7
	
	* Introduce "new" and "remove" options to allow for managing files.

	* Introduce "use_ion_aves" option to suppress (by default) calculating
	averages accross IONs the way BGL does.

	* Fixed the "no averages" code path so it rightly preserves
	the whole profile from a run.  

	* Promoted the res->transferred field to a double because as a
	32 bit quantity it could overflow for really fast or long 
	accumulations of transfers.
	
2006-01-15 Andrew Uselton <uselton2@llnl.gov>
	
	* Tag 1.6
	
	* Introduce "profiles" option, and don't do profiles unless it is
	set

	* Adjust defaults for operation without parameters file

	* Reduce verbosity for default operation, summary line only

	* Clean up "lint" detected by icc compiler, missing "returns"
	pointer arithmetic, missing "include" files <stdlib> and <string>

	* init_timer() now uses MPI_COMM_WORLD and occurs outside the 
	iteration loop.  It also announces the program name and start 
	time.

	* the fsync() seems to return errors that (probably) occurred
	earlier and were corrected.  Work around, investigate, fix.
	
2005-12-07 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.5
	
	* Discovered where the rogue ASSERT was, though not why.

	* Retrenched to only using FAIL and ASSERT and removed 
	problematic error handling code

	* Documented, but did not implement, details of system
	call return codes

	* Timer intialization check

2005-12-01 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.4
	
	* Bool enumeration

	* Optional read_only and write_only

	* Optional looping

2005-11-22 Andrew Uselton <uselton2@llnl.gov>

	* Tag 1.3

	* Code refactoring
	
2005-11-15 Andrew Uselton <uselton2@llnl.gov>
	
	* Tag 1.2

	* Progress meter

2005-11-01 Andrew Uselton <uselton2@llnl.gov>
	
	* tag 1.1
	
	* Reorganize for single directory of output

2005-10-01 Andrew Uselton <uselton2@llnl.gov>
	
	* tag 1.0
	
	* Working code

	


tools/mib.sh
#!/bin/bash
#*****************************************************************************\
#*  $Id: mib.sh 1.9 2005/11/30 20:24:57 auselton Exp $
#*****************************************************************************
#*  Copyright (C) 2001-2002 The Regents of the University of California.
#*  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
#*  Written by Andrew Uselton <uselton2@llnl.gov>
#*  UCRL-CODE-222725
#*  
#*  This file is part of Mib, an MPI-based parallel I/O benchamrk
#*  For details, see <http://www.llnl.gov/linux/mib/>.
#*  
#*  Mib is free software; you can redistribute it and/or modify it under
#*  the terms of the GNU General Public License (as published by the Free
#*  Software Foundation version 2, dated June 1991.
#*  
#*  Mib is distributed in the hope that it will be useful, but WITHOUT 
#*  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
#*  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
#*  for more details.
#*  
#*  You should have received a copy of the GNU General Public License along
#*  with Mib; if not, write to the Free Software Foundation, Inc.,
#*  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
#*****************************************************************************/
# mib.sh <srun_opts> -- <mib_opts>
# mib.sh <srun_opts> <full_path_to_mib> <mib_opts>
# 
# Run a mib test with all the logging goodies going as well.  This will create
# a timestamp labelled log directory in the $CLUSTER directory in your home 
# directory.  That log directory will have a copy of the parameters file, if 
# any, named "options".   There will be an "lwatch.log" trace of the output
#  from the lmtd server monitoring daemons (if the lwatch.py script exists).  
# There will also be a transcript of the test run in a file named "log".
# The one curve ball is that the -d option now points to the parameters file 
# rather than the log directory.  

PROG=${0##*/}
CMD_LINE="$*"
USER="auselton"
HOME="/g/g0/$USER"
BIN="$HOME/bin"
CLUSTER=
CLUSTER_BIN=

LWATCH="$BIN/lwatch.py"
HAVE_LWATCH="yes"
[ -x $LWATCH ] || HAVE_LWATCH="no"
# If the xwatch-lustre bits aren't installed you may need to include
# them locally along with this hint were to find them:
#export PYTHONPATH="$HOME/src/xwatch"

usage()
{
    [ -z $1 ] || echo "$1"
    echo "$PROG <srun_opts> -- <mib_opts>"
    echo "or"
    echo "$PROG <srun_opts><full_path_to_mib> <mib_opts>"
    echo "N.B. Use the -d arguement only to point to an optional parameters file,"
    echo "since this script will create the actual log directory."
    echo 
    exit
}

function get_date () {
    date
    DATE=`date '+%Y%m%d%H%M%S'`
}

function get_cluster () {
    CLUSTER=`nodeattr -v cluster`
    [ X"$CLUSTER" == X ] && usage "Didn't get CLUSTER: $CLUSTER"
    CLUSTER_BIN=$HOME/$CLUSTER
    MIB="$CLUSTER_BIN/mib-1.9-try"
}

function set_log () {
    LOG_DIR="$HOME/$CLUSTER/$DATE"
    mkdir -p $LOG_DIR
    LOG="$LOG_DIR/log"
    [ "$HAVE_LWATCH" == "no" ] || LWATCH_LOG="$LOG_DIR/lwatch"
# This is the first time we've been able to send anything to the log
    touch $LOG
    [ X"$LOG" != X -a -f $LOG ] || usage "Didn't get LOG: $LOG"
    echo "o DATE = $DATE" > $LOG
    echo "o CLUSTER = $CLUSTER" >> $LOG
    echo "o LOG = $LOG" >> $LOG
    chown -R $USER.$USER $LOG_DIR
    chmod go+r $LOG_DIR
}

get_date
get_cluster
set_log

# The following is usefull when there are multiple file systems
#TESTDIR=`grep testdir $CLUSTER_BIN/$PARMS | awk '{print $3}'`
#FS=${TESTDIR##/p/}
#FS=${FS%%/*}
case $CLUSTER in
    alc) FS_MONITOR="http://alci:50539"
#	[ X"$FS" == X"ga1" ] && FS_MONITOR="http://alci:50538"
	MPI_COM="srun --core=light "
	;;
    adev) FS_MONITOR="http://ilci:50538"
#	[ X"$FS" == X"ga1" ] && FS_MONITOR="http://alci:50538"
	MPI_COM="srun --core=light "
	;;
    *) usage "Edit \"mib.sh\" to list the fs monitor (lmtd) and MPI command for this cluster"
	;;
esac

# This script wants to always generate profiles and envirnmental info
HAVE_E="no"
for opt in "$*"
do
  [ "$opt" != "-E" ] || HAVE_E="true" 
#  [ "$opt" != "-p" ] || { echo "You do not want to set an explicit profile option here.  The script does it for you"; exit 1; } 
done
CMD_LINE="${CMD_LINE/--/$MIB} -p $LOG_DIR/profile"
[ "$HAVE_E" == "true" ] || CMD_LINE="$CMD_LINE -E"

# We shouldn't actually start lwatch.py until the job gets its resource,
# but that seems hard.
if [ "$HAVE_LWATCH" == "yes" ]
then
    nohup $LWATCH -f $FS_MONITOR > $LWATCH_LOG 2>&1 &
    LWATCH_PID=$!
fi

echo "$MPI_COM $CMD_LINE"
echo "tail -f $LOG"

$MPI_COM $CMD_LINE | tee -a $LOG

chown -R $USER.$USER $LOG_DIR
chmod -R go+r $LOG_DIR
rm -f $LOG~

if [ "$HAVE_LWATCH" == "yes" ]
then
    sleep 5
    kill -9 $LWATCH_PID
fi
echo "$BIN/composite.pl -d $LOG_DIR"


